x-pack/plugins/security_solution/server/endpoint/routes/actions/isolation.ts

```diff
@@ -5,81 +5,146 @@
+};
+
 /**
  * Registers the Host-(un-)isolation routes
  */
-export function registerHostIsolationRoutes(router: IRouter, endpointContext: EndpointAppContext) {
+export function registerHostIsolationRoutes(
+  router: SecuritySolutionPluginRouter,
+  endpointContext: EndpointAppContext
+) {
   // perform isolation
   router.post(
     {
-      path: `/api/endpoint/isolate`,
-      validate: {
-        body: schema.object({
-          agent_ids: schema.nullable(schema.arrayOf(schema.string())),
-          endpoint_ids: schema.nullable(schema.arrayOf(schema.string())),
-          alert_ids: schema.nullable(schema.arrayOf(schema.string())),
-          case_ids: schema.nullable(schema.arrayOf(schema.string())),
-          comment: schema.nullable(schema.string()),
-        }),
-      },
+      path: ISOLATE_HOST_ROUTE,
+      validate: HostIsolationRequestSchema,
       options: { authRequired: true },
     },
-    async (context, req, res) => {
-      if (
-        (req.body.agent_ids === null || req.body.agent_ids.length === 0) &&
-        (req.body.endpoint_ids === null || req.body.endpoint_ids.length === 0)
-      ) {
-        return res.badRequest({
-          body: {
-            message: 'At least one agent ID or endpoint ID is required',
-          },
-        });
-      }
-
-      return res.ok({
-        body: {
-          action: '713085d6-ab45-4e9e-b41d-96563cafdd97',
-        },
-      });
-    }
+    isolationRequestHandler(endpointContext, true)
   );
 
   // perform UN-isolate
   router.post(
     {
-      path: `/api/endpoint/unisolate`,
-      validate: {
-        body: schema.object({
-          agent_ids: schema.nullable(schema.arrayOf(schema.string())),
-          endpoint_ids: schema.nullable(schema.arrayOf(schema.string())),
-          alert_ids: schema.nullable(schema.arrayOf(schema.string())),
-          case_ids: schema.nullable(schema.arrayOf(schema.string())),
-          comment: schema.nullable(schema.string()),
-        }),
-      },
+      path: UNISOLATE_HOST_ROUTE,
+      validate: HostIsolationRequestSchema,
       options: { authRequired: true },
     },
-    async (context, req, res) => {
-      if (
-        (req.body.agent_ids === null || req.body.agent_ids.length === 0) &&
-        (req.body.endpoint_ids === null || req.body.endpoint_ids.length === 0)
-      ) {
-        return res.badRequest({
-          body: {
-            message: 'At least one agent ID or endpoint ID is required',
-          },
-        });
-      }
-      return res.ok({
+    isolationRequestHandler(endpointContext, false)
+  );
+}
+
+export const isolationRequestHandler = function (
+  endpointContext: EndpointAppContext,
+  isolate: boolean
+): RequestHandler<
+  unknown,
+  unknown,
+  TypeOf<typeof HostIsolationRequestSchema.body>,
+  SecuritySolutionRequestHandlerContext
+> {
+  return async (context, req, res) => {
+    if (
+      (!req.body.agent_ids || req.body.agent_ids.length === 0) &&
```

Comment:
Why are this fields `nullable` in the schema if cannot be empty? Is it because the custom message?

Reply:
They are nullable because only one or the other is required. You can send `agent_ids` and leave off `endpoint_ids`, or the other way around. I didn't look too deeply into how to achieve that with `schema`.

Reply:
Yeah, I think you can do that with [conditional](https://github.com/elastic/kibana/tree/4584a8b570402aa07832cf3e5b520e5d2cfa7166/packages/kbn-config-schema#schemaconditional)

Reply:
Started writing this as a conditional. Became illegible. Then I saw the note at the bottom of the conditional docs you linked:

> Conditional schemas may be hard to read and understand and hence should be used only sparingly.

It definitely fit that bill. I'll see if it's any easier using `io-ts`.